File: LeetCode/Java Solutions/Open the Lock/open.java
import java.util.*;

class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> deadSet = new HashSet<>(Arrays.asList(deadends));
        if (deadSet.contains("0000")) return -1;
        
        Queue<Pair<String, Integer>> queue = new LinkedList<>();
        queue.add(new Pair<>("0000", 0));
        Set<String> visited = new HashSet<>();
        visited.add("0000");
        
        while (!queue.isEmpty()) {
            Pair<String, Integer> p = queue.poll();
            String state = p.getKey();
            int depth = p.getValue();
            
            if (state.equals(target)) {
                return depth;
            }
            
            for (int i = 0; i < 4; i++) {
                int digit = state.charAt(i) - '0';
                for (int move : new int[]{-1, 1}) {
                    int nextDigit = (digit + move + 10) % 10;
                    String nextState = state.substring(0, i) + nextDigit + state.substring(i + 1);
                    
                    if (!visited.contains(nextState) && !deadSet.contains(nextState)) {
                        visited.add(nextState);
                        queue.add(new Pair<>(nextState, depth + 1));
                    }
                }
            }
        }
        
        return -1;
    }
}

File: LeetCode/Python Solutions/Open the Lock/open.py
from collections import deque
from typing import List

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        dead_set = set(deadends)
        if "0000" in dead_set:
            return -1
        
        queue = deque([("0000", 0)])  # (state, depth)
        visited = set(["0000"])
        
        while queue:
            state, depth = queue.popleft()
            
            if state == target:
                return depth
            
            for i in range(4):
                digit = int(state[i])
                # Generate the next and previous digit for each wheel
                for move in [-1, 1]:
                    next_digit = (digit + move) % 10
                    next_state = state[:i] + str(next_digit) + state[i+1:]
                    
                    if next_state not in visited and next_state not in dead_set:
                        visited.add(next_state)
                        queue.append((next_state, depth + 1))
        
        return -1
